### DISCUSSION TITLE
Documentation is very misleading regarding client-side route invalidation.

### DISCUSSION THREAD

**Discussion by Callumk7** [vercel/next.js#53019] [App Router] [answered]
Whilst they are excellent in many ways, I have been banging my head against the wall over the last couple of days trying to get my head around the client-side route cache. Imagine this fairly common use case:

1. A user navigates to their `/collection` route. This is a server component page.tsx, which fetches the data.
2. The user navigates to a `/search` route, where they find an item they want to add to their collection, they do so, updating the database in the process.
3. The user later navigates back to the `/collection` page, expecting to see their up-to-date data, but it has cached the original request, and the data is out of date.

The documentation talks at length about the fetch api, which is featured in the `/collection` page. It states that in order to force refetching, you can use properties like `revalidate: 0` or `cache: no-store`. (https://nextjs.org/docs/app/building-your-application/data-fetching/revalidating)

There are other options as well, I can try and perform on demand revalidation, I can create a route handler that uses `revalidatePath` or `revalidateTag`. 

There are also route-level config that we can use like `export const dynamic = "force-dynamic"` that looks like it should invalidate the route and force the fetch to re-occur on every request. https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config

However, this does not help our use case, as this is not what we need to invalidate, as the problem, as far as I can tell, is actually to do with routing, and 'soft-navigation'. https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#how-navigation-works

"Soft Navigation

On navigation, the cache for changed segments is reused (if it exists), and no new requests are made to the server for data."

There is also a small **good to know** on this page:
"Good to know: This client-side cache is different from the server-side [Next.js HTTP cache](https://nextjs.org/docs/app/building-your-application/data-fetching#caching-data)."

If we scroll down, we find something that looks helpful:

"Invalidating the Cache

[Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions) can be used to revalidate data on-demand by path ([revalidatePath](https://nextjs.org/docs/app/api-reference/functions/revalidatePath)) or by cache tag ([revalidateTag](https://nextjs.org/docs/app/api-reference/functions/revalidateTag))."

huh, well server actions are an alpha feature so I won't want to touch those. I had also tried using revalidatePath / revalidateTag, with route handlers instead - but this seemed to fall into the same problem as before; that request invalidation is not the same as invalidating the client-side cache. 

"Conditions for Soft Navigation

On navigation, Next.js will use soft navigation if the route you are navigating to has been [prefetched](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#prefetching), and either doesn't include [dynamic segments](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) or has the same dynamic parameters as the current route."

This is strange, because pre-fetching is disabled on the development server - so I think it is understandable to assume that no soft navigation would take place here, but that doesn't seem to be the case. Indeed setting prefetch={false} on a link back to our collection doesn't solve our problem in production either - and I guess the problem here is that even if we HAVE a dynamic route (say, `/collection/[userId]/page.tsx` ...which I also tried), is that we are navigating to **the same dynamic route**, so soft navigation is enabled, it seems. 

**The only solution I have found**
A client side `useRouter.refresh()`. This has to happen _before_ we land on the page though, as the refresh means that only the next time we visit the route will it refresh (this means we can't have a 'refresh data' button, as it won't take effect until we go away and come back again). 

We can probably have some client side useEffect to call this for us, but honestly I can't believe this is the envisiged solution. But Ive tried hard to scour the documentation, and nothing that _seems_ like it should perform like I want it too, does - so I am left with this. 

I would love to hear of what I am sure is a painfully obvious solution that I have missed - but at the same time I think its fair to be frustrated at this fairly common usage not acting in a way that seems logical, based on the documentation provided.

**Comment by axelvaindal** [comment]
@Callumk7 I've encountered the same thing and it's kinda weird indeed.

Regarding your point here:

> /collection/[userId]/page.tsx ...which I also tried), is that we are navigating to the same dynamic route, so soft navigation is enabled, it seems.

Yes, because the route segment `userId` has not changed, so Next fetches the page from the cache, which have the old data and not the new one.
I've got the same issue myself when using `a` and `a/create` where the second redirects to the first one.

In my case, using `router.refresh()` seems to solve the issue as I redirect after saving, so it just invalidate the cache, redirect, fetch the new data and it works fine.

I've no idea if this is the best practice for such common use case, so I'd be happy if anyone from Vercel can step in here and provide guidance.

@leerob @steven-tey Can you provide guidance regarding soft-navigation / hard-navigation use-cases ? 🙏

**Reply by leerob** [reply]
https://github.com/vercel/next.js/pull/52514

**Comment by leerob** [comment, marked as answer]
This is being addressed in https://github.com/vercel/next.js/pull/52514, please give that a read! Thank you.

**Reply by Callumk7** [reply]
Thats awesome, great to see
